File: vb/demo/Ice/bidir/CallbackSenderI.vb
' **********************************************************************
'
' Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
'
' This copy of Ice is licensed to you under the terms described in the
' ICE_LICENSE file included in this distribution.
'
' **********************************************************************

Imports BidirDemo
Imports System.Collections

Class CallbackSenderI
    Inherits CallbackSenderDisp_

    Public Sub New(ByVal communicator As Ice.Communicator)
        _communicator = communicator
        _destroy = False
        _clients = New ArrayList
    End Sub

    Public Sub destroy()
        SyncLock Me
            System.Console.Out.WriteLine("destroying callback sender")
            _destroy = True

            System.Threading.Monitor.Pulse(Me)
        End SyncLock
    End Sub

    Public Overloads Overrides Sub addClient(ByVal ident As Ice.Identity, ByVal current As Ice.Current)
        SyncLock Me
            System.Console.Out.WriteLine("adding client `" & current.adapter.getCommunicator().identityToString(ident) & "'")

            Dim base As Ice.ObjectPrx = current.con.createProxy(ident)
            Dim client As CallbackReceiverPrx = CallbackReceiverPrxHelper.uncheckedCast(base)
            _clients.Add(client)
        End SyncLock
    End Sub

    Public Sub Run()
        Dim num As Integer = 0
        While True
            Dim clients As ArrayList
            SyncLock Me
                System.Threading.Monitor.Wait(Me, 2000)
                If _destroy Then
                    Exit While
                End If
                clients = New ArrayList(_clients)
            End SyncLock

            If clients.Count > 0 Then
                num += 1

                For Each c As CallbackReceiverPrx In clients
                    Try
                        c.callback(num)
                    Catch ex As Ice.LocalException
                        System.Console.Out.WriteLine("removing client `" & _communicator.identityToString(c.ice_getIdentity()) & "'")
                        SyncLock Me
                            _clients.Remove(c)
                        End SyncLock
                    End Try
                Next
            End If
        End While
    End Sub

    Private _communicator As Ice.Communicator
    Private _destroy As Boolean
    Private _clients As ArrayList

End Class

File: cs/demo/Ice/compact/bidir/CallbackSenderI.cs
// **********************************************************************
//
// Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

using Demo;
using System;
using System.Collections;

class CallbackSenderI : CallbackSenderDisp_
{
    public CallbackSenderI(Ice.Communicator communicator)
    {
        _communicator = communicator;
        _destroy = false;
        _clients = new ArrayList();
    }

    public void destroy()
    {
        lock(this)
        {
            System.Console.Out.WriteLine("destroying callback sender");
            _destroy = true;
            
            System.Threading.Monitor.Pulse(this);
        }
    }

    public override void addClient(Ice.Identity ident, Ice.Current current)
    {
        lock(this)
        {
            System.Console.Out.WriteLine("adding client `" + _communicator.identityToString(ident) + "'");

            Ice.ObjectPrx @base = current.con.createProxy(ident);
            CallbackReceiverPrx client = CallbackReceiverPrxHelper.uncheckedCast(@base);
            _clients.Add(client);
        }
    }

    public void Run()
    {
        int num = 0;
        while(true)
        {
            ArrayList clients;
            lock(this)
            {
                System.Threading.Monitor.Wait(this, 2000);
                if(_destroy)
                {
                    break;
                }

                clients = new ArrayList(_clients);
            }

            if(clients.Count > 0)
            {
                ++num;
                foreach(CallbackReceiverPrx c in clients)
                {
                    try
                    {
                        c.callback(num);
                    }
                    catch(Ice.LocalException ex)
                    {
                        Console.Error.WriteLine("removing client `" +
                                                _communicator.identityToString(c.ice_getIdentity()) + "':\n" + ex);
                        lock(this)
                        {
                            _clients.Remove(c);
                        }
                    }
                }
            }
        }
    }

    private Ice.Communicator _communicator;
    private bool _destroy;
    private ArrayList _clients;
}
